### PR TITLE
Improve Online Learning performance

### LINKED ISSUE #18: Performance problems in online learning
**Describe the bug**
Online learning is currently slow - taking large amounts of time to process one step. It seems that a single processor is busy. This could be due to:
- Merging and copying of xarray.datasets.
- Predict/Transform of single datapoints by sklearn.
- Internal calls of pyWATTS

**Expected behavior**
The online learning should run smoothly. Posisble solutions:
- Reduce copying and merging of datasets.
- Multiprocessing.

### PR DESCRIPTION
**Pull request by benHeid**
## Description
Improved online learning performance

## Related Issues
closes #18 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos)
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Did you update the CHANGELOG

## PR review
Anyone in the community is free to review the PR once the tests have passed.

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones to the PR so it can be classified

### CODE REVIEW

File: pywatts/core/base_step.py [hunk 1]
```diff
@@ -109,14 +116,15 @@ def _pack_data(self, start, end, buffer_element=None, return_all=False):
...
-                return self.buffer.copy()[buffer_element].sel(
+                return self.buffer[buffer_element].sel(
                     **{time_index[0]: index[(index >= start) & (index < end.to_numpy())]})
             elif return_all:
-                return {key: b.copy().sel(**{time_index[0]: index[(index >= start) & (index < end.to_numpy())]}) for
+                return {key: b.sel(**{time_index[0]: index[(index >= start) & (index < end.to_numpy())]}) for
                         key, b in self.buffer.items()}
             else:
-                return list(self.buffer.copy().values())[0].sel(
+                return list(self.buffer.values())[0].sel(
```
[the remark this answers is not in the document]
Comment (olineumann) [reply pywatts/core/base_step.py:127]: Yes just checked it. Not working correctly IMO.

```python
import numpy as np
import pandas as pd

from pywatts.core.pipeline import Pipeline
from pywatts.callbacks import PrintCallback
from pywatts.wrapper import FunctionModule


def method_a(x):
    x[0] = 1
    return x


def method_b(x):
    x[1] = 2
    return x


def main():
    pipeline = Pipeline('run')
    a = FunctionModule(method_a, name='FA')(x=pipeline['x'], callbacks=[PrintCallback()])
    b = FunctionModule(method_b, name='FB')(x=pipeline['x'], callbacks=[PrintCallback()])

    df = pd.DataFrame()
    df['x'] = np.array([0, 0, 0, 0])
    df.index = pd.date_range("2018-01-01", periods=4, freq="H")
    pipeline.train(df)


if __name__ == '__main__':
    main()
```
Output:
```

# Print Callback
## FA
index
2018-01-01 00:00:00    1
2018-01-01 01:00:00    0
2018-01-01 02:00:00    0
2018-01-01 03:00:00    0
Freq: H, dtype: int64

# Print Callback
## FB
index
2018-01-01 00:00:00    1
2018-01-01 01:00:00    2
2018-01-01 02:00:00    0
2018-01-01 03:00:00    0
```

**Review by olineumann — CHANGES_REQUESTED** [review]
Need to check if a buffer is used by multiple modules afterwards.

File: pywatts/core/base_step.py [hunk 2]
```diff
@@ -62,25 +63,31 @@ def get_result(self, start: pd.Timestamp, end: Optional[pd.Timestamp], buffer_el
...
+                self.cached_result = self._compute(start, end), start, end
                 self._current_end = end
             if not end:
                 self.finished = True
             else:
                 self.finished = not self.further_elements(end)
-            self._callbacks()
 
+            # Only call callbacks if the step is finished
+            if self.finished:
```
[the remark this answers is not in the document]
Comment (benHeid) [reply pywatts/core/base_step.py:81]: I don't think so. Because if the user do not want to have callbacks, he/she should not add callbacks to the step/module.
